File: 19-Remove Nth Node From End of List-删除链表的倒数第 N 个结点/19-Remove Nth Node From End of List-删除链表的倒数第 N 个结点.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode virtualHead = new ListNode(0, head);
        ListNode pre;
        ListNode p = virtualHead;
        while(n-- > 0){
            p = p.next;
        }

        pre = virtualHead;
        while(p.next != null){
            pre = pre.next;
            p = p.next;
        }
        if(pre.next != null){
            pre.next = pre.next.next;
        }
        return virtualHead.next;
    }
}

File: 19-Remove Nth Node From End of List-删除链表的倒数第 N 个结点/19-Remove Nth Node From End of List-删除链表的倒数第 N 个结点.js

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
 var removeNthFromEnd = function(head, n) {
    let virtualHead = new ListNode(0, head)
    let count = n 
    let p = virtualHead
    let targetParent = virtualHead
    while(p != null){
        p = p.next
        if(count > 0){
            count--
        }else if(count === 0){
            //这一步p移动了，但是targetParent没有移动，所以其间隔加了1，所以可以直接next，这样算出来的targetParent就是要删除节点的pre节点，而不是要删除的节点
            targetParent = virtualHead
            count = -1
        }else{
            targetParent = targetParent.next
        }
    }
    

    if(targetParent.next){
        targetParent.next =targetParent.next.next
    }
    
    return virtualHead.next
};